Extend MultipartBodyBuilder methods so they accept a filename parameter

koziolk opened the pull request:
This fix allows passing a filename to MultipartBodyBuilder methods and
sets the proper value of the content disposition header based on it.

Closes gh-22616

koziolk committed:
Extend MultipartBodyBuilder methods so they accept a filename parameter

This fix allows passing a filename to MultipartBodyBuilder methods and
sets the proper value of the content disposition header based on it.

Closes gh-22616

koziolk committed:
Extend MultipartBodyBuilder methods so they accept a filename parameter

This fix allows passing a filename to MultipartBodyBuilder methods and
sets the proper value of the content disposition header based on it.

Closes gh-22616

ielatif commented on spring-web/src/main/java/org/springframework/http/client/MultipartBodyBuilder.java:
```diff
@@ -100,6 +124,10 @@ public PartBuilder part(String name, Object part, @Nullable MediaType contentTyp
 		Object partBody;
 		HttpHeaders partHeaders = new HttpHeaders();
 
+		if (filename != null) {
```
You should also check that filename has length.
if (StringUtils.hasLength(filename))

koziolk replied:
Right, fixing in my next change.

ielatif commented on spring-web/src/main/java/org/springframework/http/client/MultipartBodyBuilder.java:
```diff
@@ -137,6 +165,30 @@ public PartBuilder part(String name, Object part, @Nullable MediaType contentTyp
+	 * @return builder that allows for further customization of part headers
+	 */
+	public <T, P extends Publisher<T>> PartBuilder asyncPart(
+			String name, String filename, P publisher, Class<T> elementClass) {
```
filename should be optional.

koziolk committed:
Extend MultipartBodyBuilder methods so they accept a filename parameter

This fix allows passing a filename to MultipartBodyBuilder methods and
sets the proper value of the content disposition header based on it.

Closes gh-22616

rstoyanchev commented:
I'm not sure this is headed in the right direction, it doubles the number of methods. I would add a single dedicated method for file parts, something like:

```java
public PartBuilder filePart(String name, String filename, Publishder<DataBuffer> content);
```

Also more generally, please organize into logical commits via squash. In this PR I would expect one commit only.

rstoyanchev commented:
I'll resolve this differently in #23083 through a filename parameter on `PartBuilder`.
